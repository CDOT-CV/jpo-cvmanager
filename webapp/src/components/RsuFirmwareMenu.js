import React from 'react'
import { confirmAlert } from 'react-confirm-alert'
import 'react-confirm-alert/src/react-confirm-alert.css'
import { useDispatch, useSelector } from 'react-redux'

import {
  selectFirmwareUpgradeAvailable,
  selectFirmwareUpgradeName,
  selectFirmwareUpgradeMsg,
  selectFirmwareUpgradeErr,

  // Actions
  checkFirmwareUpgrade,
  startFirmwareUpgrade,
} from '../generalSlices/configSlice'

import { selectRsuIpv4 } from '../generalSlices/rsuSlice'

import './css/SnmpwalkMenu.css'

const RsuFirmwareMenu = () => {
  const dispatch = useDispatch()
  const firmwareUpgradeAvailable = useSelector(selectFirmwareUpgradeAvailable)
  const firmwareUpgradeName = useSelector(selectFirmwareUpgradeName)
  const firmwareUpgradeMsg = useSelector(selectFirmwareUpgradeMsg)
  const firmwareUpgradeErr = useSelector(selectFirmwareUpgradeErr)

  const rsuIp = useSelector(selectRsuIpv4)

  const options = {
    title: 'RSU Firmware Upgrade',
    message:
      'Are you sure you would like to run this firmware upgrade? This can take up to an extended period of time to complete: 5 - 60 minutes. Once an upgrade has started, you may continue to use or close the CV Manager web application without interfering with the upgrade.',
    buttons: [
      {
        label: 'Yes',
        onClick: () => dispatch(startFirmwareUpgrade([rsuIp])),
      },
      {
        label: 'No',
        onClick: () => {},
      },
    ],
    childrenElement: () => <div />,
    closeOnEscape: true,
    closeOnClickOutside: true,
    keyCodeForClose: [8, 32],
    willUnmount: () => {},
    afterClose: () => {},
    onClickOutside: () => {},
    onKeypressEscape: () => {},
  }

  return (
    <div id="snmpdiv">
      <h2 class="firmwareHeader">Firmware Upgrade</h2>

      {firmwareUpgradeAvailable ? (
        <div>
          <div id="firmwarediv">
            <p id="firmwarenoticetext">A firmware upgrade is available!</p>
            <p id="firmwaresecondarytext">
              <b>Version: {firmwareUpgradeName}</b>
            </p>
          </div>

          <button id="refreshbtn" onClick={() => confirmAlert(options)}>
            Run Firmware Upgrade
          </button>
        </div>
      ) : (
        <div>
          <div id="firmwarediv">
            <p id="firmwaretext">Check for the latest available RSU firmware upgrades and install them</p>
            {firmwareUpgradeMsg !== '' && (
              <div>
                {firmwareUpgradeErr ? (
                  <p id="warningtext">{firmwareUpgradeMsg}</p>
                ) : (
                  <p id="successtext">{firmwareUpgradeMsg}</p>
                )}
              </div>
            )}
          </div>

          <button id="refreshbtn" onClick={() => dispatch(checkFirmwareUpgrade([rsuIp]))}>
            Check For Upgrade Availability
          </button>
        </div>
      )}
    </div>
  )
}

export default RsuFirmwareMenu
