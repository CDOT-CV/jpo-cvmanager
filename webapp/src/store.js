import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
