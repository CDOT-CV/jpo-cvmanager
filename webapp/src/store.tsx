import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
