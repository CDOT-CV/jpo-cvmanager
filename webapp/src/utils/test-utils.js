/**
 * This function searches for the every react-aria SSR ids in a given HTMLElement node and replace every attribute values with a static id
 *
 * This can be usefull when you're trying to generate a snapshot of components using react-aria under the hood
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceReactAriaIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container) {
  const props = [
    {
      selector: "class",
      updateFunc: (val) =>
        val.replace(
          /css-[0-9a-z]{6}-MuiTableCell-root-MuiTablePagination-root/g,
          "css-mocked-MuiTableCell-root-MuiTablePagination-root"
        ),
    },
  ];

  container.querySelectorAll("td").forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)));
      }
    });
  });
  container.querySelectorAll("input[aria-invalid]").forEach((item) => {
    delete item["aria-invalid"];
  });
  return container;
}

export { replaceChaoticIds };
