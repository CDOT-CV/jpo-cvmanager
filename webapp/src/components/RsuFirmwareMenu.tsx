import React from 'react'
import { confirmAlert } from 'react-confirm-alert'
import 'react-confirm-alert/src/react-confirm-alert.css'
import { useDispatch, useSelector } from 'react-redux'

import {
  selectFirmwareUpgradeAvailable,
  selectFirmwareUpgradeName,
  selectFirmwareUpgradeMsg,
  selectFirmwareUpgradeErr,

  // Actions
  checkFirmwareUpgrade,
  startFirmwareUpgrade,
} from '../generalSlices/configSlice'

import './css/SnmpwalkMenu.css'
import { AnyAction, ThunkDispatch } from '@reduxjs/toolkit'
import { RootState } from '../store'

interface RsuFirmwareMenuProps {
  type: string
  rsuIpList: string[]
}

const RsuFirmwareMenu = (props: RsuFirmwareMenuProps) => {
  const dispatch: ThunkDispatch<RootState, void, AnyAction> = useDispatch()
  const firmwareUpgradeAvailable = useSelector(selectFirmwareUpgradeAvailable)
  const firmwareUpgradeName = useSelector(selectFirmwareUpgradeName)
  const firmwareUpgradeMsg = useSelector(selectFirmwareUpgradeMsg)
  const firmwareUpgradeErr = useSelector(selectFirmwareUpgradeErr)

  const options = {
    title: 'RSU Firmware Upgrade',
    message:
      'Are you sure you would like to run a firmware upgrade? This can take up to an extended period of time to complete: 5 - 60 minutes. Once an upgrade has started, you may continue to use or close the CV Manager web application without interfering with the upgrade.',
    buttons: [
      {
        label: 'Yes',
        onClick: () => dispatch(startFirmwareUpgrade(props.rsuIpList)),
      },
      {
        label: 'No',
        onClick: () => {},
      },
    ],
    childrenElement: () => <div />,
    closeOnEscape: true,
    closeOnClickOutside: true,
    keyCodeForClose: [8, 32],
    willUnmount: () => {},
    afterClose: () => {},
    onClickOutside: () => {},
    onKeypressEscape: () => {},
  }

  return (
    <div id="snmpdiv">
      <h2 className="firmwareHeader">Firmware Upgrade</h2>

      {props.type === 'single_rsu' && (
        <div>
          {firmwareUpgradeAvailable ? (
            <div>
              <div id="firmwarediv">
                <p id="firmwarenoticetext" role="status">
                  A firmware upgrade is available!
                </p>
                <p id="firmwaresecondarytext">
                  <b>Version: {firmwareUpgradeName}</b>
                </p>
              </div>

              <button id="refreshbtn" onClick={() => confirmAlert(options)}>
                Run Firmware Upgrade
              </button>
            </div>
          ) : (
            <div>
              <div id="firmwarediv">
                <p id="firmwaretext">Check for the latest available RSU firmware upgrades and install them</p>
                {firmwareUpgradeMsg !== '' && (
                  <div>
                    {firmwareUpgradeErr ? (
                      <p id="warningtext" role="alert">
                        {firmwareUpgradeMsg}
                      </p>
                    ) : (
                      <p id="successtext" role="status">
                        {firmwareUpgradeMsg}
                      </p>
                    )}
                  </div>
                )}
              </div>

              <button id="refreshbtn" onClick={() => dispatch(checkFirmwareUpgrade(props.rsuIpList))}>
                Check For Upgrade Availability
              </button>
            </div>
          )}
        </div>
      )}

      {props.type === 'multi_rsu' && (
        <div>
          <div id="firmwarediv">
            <p id="firmwaretext">
              Submit all selected RSUs to have their firmware upgraded if available. RSUs that are already up to date
              will be skipped. If the RSU is offline, it will be marked for an upgrade and will be upgraded when it
              comes back online.
            </p>
          </div>

          <button id="refreshbtn" onClick={() => confirmAlert(options)}>
            Run Firmware Upgrades
          </button>
        </div>
      )}
    </div>
  )
}

export default RsuFirmwareMenu
