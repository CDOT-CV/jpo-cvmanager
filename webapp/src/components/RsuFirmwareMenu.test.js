import React from 'react'
import { render } from '@testing-library/react'
import RsuFirmwareMenu from './RsuFirmwareMenu'
import { Provider } from 'react-redux'
import { setupStore } from '../store'
import { replaceChaoticIds } from '../utils/test-utils'

it('should take a snapshot', () => {
  const { container } = render(
    <Provider store={setupStore({})}>
      <RsuFirmwareMenu />
    </Provider>
  )

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
