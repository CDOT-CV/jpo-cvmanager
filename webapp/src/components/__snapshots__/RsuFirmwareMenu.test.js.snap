// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div
    id="snmpdiv"
  >
    <h2
      class="firmwareHeader"
    >
      Firmware Upgrade
    </h2>
    <div>
      <div
        id="firmwarediv"
      >
        <p
          id="firmwaretext"
        >
          Check for the latest available RSU firmware upgrades and install them
        </p>
      </div>
      <button
        id="refreshbtn"
      >
        Check For Upgrade Availability
      </button>
    </div>
  </div>
</div>
`;
